Extend MatchedDeclarationsCache to store multiple entires

Andrew-Fryer opened the pull request:
#### 484daed3f84dda2c6b208eb27552f45fa3f883b0
<pre>
Extend MatchedDeclarationsCache to store multiple entires
<a href="https://bugs.webkit.org/show_bug.cgi?id=291829">https://bugs.webkit.org/show_bug.cgi?id=291829</a>
<a href="https://rdar.apple.com/149652557">rdar://149652557</a>

Reviewed by Antti Koivisto.

Extend that hash table by using wtf Vector of Entry
as the value type instead of just Entry.

* Source/WebCore/style/MatchedDeclarationsCache.cpp:
(WebCore::Style::MatchedDeclarationsCache::find):
(WebCore::Style::MatchedDeclarationsCache::add):
(WebCore::Style::MatchedDeclarationsCache::num_entries):
(WebCore::Style::MatchedDeclarationsCache::clearEntriesAffectedByViewportUnits):
(WebCore::Style::MatchedDeclarationsCache::sweep):
* Source/WebCore/style/MatchedDeclarationsCache.h:
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::applyMatchedProperties):

Canonical link: <a href="https://commits.webkit.org/294198@main">https://commits.webkit.org/294198@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0289723f7a437b24fda7aa3dbee8fc46baa5df03

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100969 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20631 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10934 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106115 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51593 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103010 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20940 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29125 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76893 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33918 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103976 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16109 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91198 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57241 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15924 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9303 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50943 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85830 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9281 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108471 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28097 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20665 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85857 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28459 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87398 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85399 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21765 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30115 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7842 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22140 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28027 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33295 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27839 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31159 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29397 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Andrew-Fryer commented on Source/WebCore/style/MatchedDeclarationsCache.cpp:
```diff
@@ -169,7 +178,18 @@ void MatchedDeclarationsCache::add(const RenderStyle& style, const RenderStyle&
+    auto it = m_entries.find(hash);
+    if (it != m_entries.end()) {
+        if (it->value.size() < maxEntriesPerHash)
+            it->value.append(Entry { &matchResult, RenderStyle::clonePtr(style), RenderStyle::clonePtr(parentStyle), userAgentAppearanceStyleCopy() });
```
I'd like to `constructAndAppend` here (and in the else case), but EWS has an older compiler that complained:
https://ews-build.webkit.org/#/builders/138/builds/10322/steps/14/logs/stdio

Andrew-Fryer replied:
It seems that we need clang 17 (rather than 16), which requires an update to the bots, so I'll come back to this later.

anttijk commented on Source/WebCore/style/MatchedDeclarationsCache.cpp:
```diff
@@ -133,7 +133,7 @@ unsigned MatchedDeclarationsCache::computeHash(const MatchResult& matchResult, c
 }
 
-const MatchedDeclarationsCache::Entry* MatchedDeclarationsCache::find(unsigned hash, const MatchResult& matchResult, const StyleCustomPropertyData& inheritedCustomProperties)
+const MatchedDeclarationsCache::Entry* MatchedDeclarationsCache::find(unsigned hash, const MatchResult& matchResult, const StyleCustomPropertyData& inheritedCustomProperties, const RenderStyle& parentStyle, bool& inheritedEqual)
```
It would be nicer to not use a return value parameter. Instead you can for example add
```
struct Result {
   MatchedDeclarationsCache& entry;
   bool inheritedEqual;
}
```

in `MatchedDeclarationsCache` scope and return it as `std::optional<Result> find(...)`

Andrew-Fryer replied:
Great, I'll do this!
(I actually originally wrote using `std::pair`, but thought that was too awkward.)

anttijk replied:
Yeah, structs are convenient in modern C++ and read better than std::pair.

anttijk commented on Source/WebCore/style/MatchedDeclarationsCache.cpp:
```diff
@@ -169,7 +178,18 @@ void MatchedDeclarationsCache::add(const RenderStyle& style, const RenderStyle&
+        newBucket.reserveCapacity(maxEntriesPerHash);
+        newBucket.append(Entry { &matchResult, RenderStyle::clonePtr(style), RenderStyle::clonePtr(parentStyle), userAgentAppearanceStyleCopy() });
+        m_entries.add(hash, WTFMove(newBucket));
+    }
```
Please rewrite this using `HashMap::ensure()`.

"Store `Entry`s out of line because hash table complains..." is not a good code comment. There is no need to explain why you are not doing things in a wrong way. 

Another option here would be to use fixed size `std::array<std::unique_ptr<Entry>, 4>` with null entry indicating an empty slot. 

For future work, it would be nicer if there was a cache replacement strategy. If new styles keep getting added we just drop them if there is no space. Some of the entries may be quite useless and would be better replaced.

anttijk commented on Source/WebCore/style/MatchedDeclarationsCache.cpp:
```diff
@@ -186,8 +206,13 @@ void MatchedDeclarationsCache::clearEntriesAffectedByViewportUnits()
     Ref protectedThis { *this };
 
+    for (auto& [key, bucket] : m_entries) {
+        bucket.removeAllMatching([&] (const Entry& entry) -> bool {
```
Since there is another client you could add a `MatchedDeclarationsCache::removeIf` helper that takes a function.

anttijk commented on Source/WebCore/style/MatchedDeclarationsCache.cpp:
```diff
@@ -186,8 +206,13 @@ void MatchedDeclarationsCache::clearEntriesAffectedByViewportUnits()
     m_entries.removeIf([](auto& keyValue) {
-        return keyValue.value.renderStyle->usesViewportUnits();
+        return !keyValue.value.size();
     });
```
it would also take care of this.

Andrew-Fryer commented on Source/WebCore/style/MatchedDeclarationsCache.cpp:
```diff
@@ -169,7 +177,14 @@ void MatchedDeclarationsCache::add(const RenderStyle& style, const RenderStyle&
+        newBucket.reserveCapacity(maxEntriesPerHash);
+        return newBucket;
+    });
+    if (addResult.iterator->value.size() < maxEntriesPerHash)
```
I don't think I need to `ASSERT(addResult.iterator.get());` here because all the calls to the `AddResult` constructor I see in HashTable.h use `makeKnownGoodIterator` (and conceptually the iterator should have a value), but maybe I should have an assertion here to be sure...?

anttijk commented on Source/WebCore/style/MatchedDeclarationsCache.cpp:
```diff
@@ -169,7 +177,14 @@ void MatchedDeclarationsCache::add(const RenderStyle& style, const RenderStyle&
-    m_entries.add(hash, Entry { &matchResult, RenderStyle::clonePtr(style), RenderStyle::clonePtr(parentStyle), userAgentAppearanceStyleCopy() });
+    constexpr unsigned maxEntriesPerHash = 4;
+    auto addResult = m_entries.ensure(hash, [&] {
+        Vector<Entry> newBucket;
+        newBucket.reserveCapacity(maxEntriesPerHash);
+        return newBucket;
```
It is kinda wasteful in terms of memory to reserve space for 4 entries when if there is often only one (though this is still better than not reserving capacity at all due to Vector growth strategy). Something like `std::array<std::unique_ptr<Entry>, 4>` might be more optimal. But this can be tweaked later.

anttijk commented on Source/WebCore/style/StyleResolver.cpp:
```diff
@@ -714,32 +714,32 @@ void Resolver::applyMatchedProperties(State& state, const MatchResult& matchResu
-    auto hasUsableEntry = cacheEntry && MatchedDeclarationsCache::isCacheable(element, style, parentStyle);
+    auto hasUsableEntry = cacheResult && MatchedDeclarationsCache::isCacheable(element, style, parentStyle);
     if (hasUsableEntry) {
+        const MatchedDeclarationsCache::Entry& cacheEntry = cacheResult->entry;
```
I'd just say `auto`

anttijk commented on Source/WebCore/style/MatchedDeclarationsCache.cpp:
```diff
@@ -133,23 +133,31 @@ unsigned MatchedDeclarationsCache::computeHash(const MatchResult& matchResult, c
+
+    if (result)
+        return Result { *result, false };
+    return std::nullopt;
```
I'd reverse these. It is nicer if the success flow goes to the end of the function and fail cases bail out. You could use named initializers for clarity

`return Result { .entry = *result, .inheritedEqual = false };`

anttijk commented on Source/WebCore/style/MatchedDeclarationsCache.cpp:
```diff
@@ -133,23 +133,31 @@ unsigned MatchedDeclarationsCache::computeHash(const MatchResult& matchResult, c
-    auto& entry = it->value;
-    if (!matchResult.cacheablePropertiesEqual(*entry.matchResult))
-        return nullptr;
+    Entry* result = nullptr;
```
It would be nicer if the type of a variable called `result` was `Result` or `std::optional<Result>` since we have a type of that name here. I think you can also simplify the flow a bit by using the latter since you can just return it at the end.

Andrew-Fryer committed:
Extend MatchedDeclarationsCache to store multiple entires
https://bugs.webkit.org/show_bug.cgi?id=291829
rdar://149652557

Reviewed by Antti Koivisto.

Extend that hash table by using wtf Vector of Entry
as the value type instead of just Entry.

* Source/WebCore/style/MatchedDeclarationsCache.cpp:
(WebCore::Style::MatchedDeclarationsCache::find):
(WebCore::Style::MatchedDeclarationsCache::add):
(WebCore::Style::MatchedDeclarationsCache::num_entries):
(WebCore::Style::MatchedDeclarationsCache::clearEntriesAffectedByViewportUnits):
(WebCore::Style::MatchedDeclarationsCache::sweep):
* Source/WebCore/style/MatchedDeclarationsCache.h:
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::applyMatchedProperties):

Canonical link: https://commits.webkit.org/294198@main
